File: TDA-Vehiculos/Java/Vehiculo.java
// CONCEPTO: CLASE ABSTRACTA
// Contiene los atributos y comportamientos comunes de los vehículos.
public abstract class Vehiculo {

    // CONCEPTO: ENCAPSULAMIENTO
    // protected permite que las clases hijas utilicen estos atributos.
    protected String placa;
    protected String marca;
    protected String modelo;

    // CONCEPTO: TIPOS DE DATOS PRIMITIVOS
    protected int anio;
    protected double precio;
    protected boolean disponible;

    // CONCEPTO: CONSTRUCTOR
    // Inicializa los atributos de un vehículo.
    public Vehiculo(String placa, String marca, String modelo,
                    int anio, double precio, boolean disponible) {

        if (placa == null || placa.trim().isEmpty()) {
            throw new IllegalArgumentException(
                    "La placa no puede estar vacía.");
        }

        if (marca == null || marca.trim().isEmpty()) {
            throw new IllegalArgumentException(
                    "La marca no puede estar vacía.");
        }

        if (modelo == null || modelo.trim().isEmpty()) {
            throw new IllegalArgumentException(
                    "El modelo no puede estar vacío.");
        }

        if (anio < 1886 || anio > 2100) {
            throw new IllegalArgumentException(
                    "El año debe estar entre 1886 y 2100.");
        }

        if (precio < 0) {
            throw new IllegalArgumentException(
                    "El precio no puede ser negativo.");
        }

        this.placa = placa.trim().toUpperCase();
        this.marca = marca.trim();
        this.modelo = modelo.trim();
        this.anio = anio;
        this.precio = precio;
        this.disponible = disponible;
    }

    // Permite obtener la placa desde el TDA RegistroVehiculos.
    public String getPlaca() {
        return placa;
    }

    // Muestra los atributos comunes.
    protected void mostrarDatosComunes() {
        System.out.println("Placa: " + placa);
        System.out.println("Marca: " + marca);
        System.out.println("Modelo: " + modelo);
        System.out.println("Año: " + anio);
        System.out.printf("Precio: $%.2f%n", precio);
        System.out.println(
                "Disponible: " + (disponible ? "Sí" : "No"));
    }

    // CONCEPTO: MÉTODO ABSTRACTO
    // Las clases hijas implementarán este método de manera diferente.
    public abstract void mostrarInformacion();
}

File: TDA-Vehiculos-c++/Vehiculo.cpp
#include "Vehiculo.h"
#include <iostream>
#include <iomanip>
#include <stdexcept>
#include <algorithm>
#include <cctype>

using namespace std;

// Funciones auxiliares para emular el trim() y toUpperCase() de Java
string trim(const string& str) {
    size_t first = str.find_first_not_of(" \t\r\n");
    if (string::npos == first) return "";
    size_t last = str.find_last_not_of(" \t\r\n");
    return str.substr(first, (last - first + 1));
}

string toUpperCase(string str) {
    transform(str.begin(), str.end(), str.begin(), ::toupper);
    return str;
}

// CONCEPTO: CONSTRUCTOR
// Inicializa los atributos de un vehículo.
Vehiculo::Vehiculo(string placa, string marca, string modelo,
                   int anio, double precio, bool disponible) {

    string placaTrim = trim(placa);
    if (placaTrim.empty()) {
        throw invalid_argument("La placa no puede estar vacía.");
    }

    string marcaTrim = trim(marca);
    if (marcaTrim.empty()) {
        throw invalid_argument("La marca no puede estar vacía.");
    }

    string modeloTrim = trim(modelo);
    if (modeloTrim.empty()) {
        throw invalid_argument("El modelo no puede estar vacío.");
    }

    if (anio < 1886 || anio > 2100) {
        throw invalid_argument("El año debe estar entre 1886 y 2100.");
    }

    if (precio < 0) {
        throw invalid_argument("El precio no puede ser negativo.");
    }

    this->placa = toUpperCase(placaTrim);
    this->marca = marcaTrim;
    this->modelo = modeloTrim;
    this->anio = anio;
    this->precio = precio;
    this->disponible = disponible;
}

// Permite obtener la placa desde el TDA RegistroVehiculos.
string Vehiculo::getPlaca() const {
    return placa;
}

// Muestra los atributos comunes.
void Vehiculo::mostrarDatosComunes() const {
    cout << "Placa: " << placa << endl;
    cout << "Marca: " << marca << endl;
    cout << "Modelo: " << modelo << endl;
    cout << "Año: " << anio << endl;
    // std::fixed y std::setprecision(2) emulan el comportamiento de printf("%.2f")
    cout << "Precio: $" << fixed << setprecision(2) << precio << endl;
    cout << "Disponible: " << (disponible ? "Sí" : "No") << endl;
}